cpp/src/community/matching_impl.cuh

```diff
@@ -0,0 +1,382 @@
+        cugraph::edge_dst_dummy_property_t{}.view(),
+        edge_weight_view,
+        graph_view_t::is_multi_gpu
+          ? src_key_cache.view()
+          : detail::edge_major_property_view_t<vertex_t, vertex_t const*>(local_vertices.begin()),
+        [] __device__(auto, auto dst, thrust::nullopt_t, thrust::nullopt_t, auto wt) {
+          return thrust::make_tuple(wt, dst);
+        },
+        thrust::make_tuple(weight_t{0.0}, invalid_partner),
+        reduce_op::maximum<thrust::tuple<weight_t, vertex_t>>{},
+        true);
+
+    //
+    // For each target, find the best offer
+    //
+
+    if constexpr (graph_view_t::is_multi_gpu) {
+      auto vertex_partition_range_lasts = current_graph_view.vertex_partition_range_lasts();
+
+      rmm::device_uvector<vertex_t> d_vertex_partition_range_lasts(
+        vertex_partition_range_lasts.size(), handle.get_stream());
+
+      raft::update_device(d_vertex_partition_range_lasts.data(),
+                          vertex_partition_range_lasts.data(),
+                          vertex_partition_range_lasts.size(),
+                          handle.get_stream());
+
+      auto& major_comm = handle.get_subcomm(cugraph::partition_manager::major_comm_name());
+      auto const major_comm_size = major_comm.get_size();
+      auto& minor_comm = handle.get_subcomm(cugraph::partition_manager::minor_comm_name());
+      auto const minor_comm_size = minor_comm.get_size();
+
+      auto func = cugraph::detail::compute_gpu_id_from_int_vertex_t<vertex_t>{
+        raft::device_span<vertex_t const>(d_vertex_partition_range_lasts.data(),
+                                          d_vertex_partition_range_lasts.size()),
+        major_comm_size,
+        minor_comm_size};
+
+      rmm::device_uvector<size_t> d_tx_value_counts(0, handle.get_stream());
+
+      auto triplet_first = thrust::make_zip_iterator(
+        candidates.begin(), offers_from_candidates.begin(), targets.begin());
+
+      d_tx_value_counts = cugraph::groupby_and_count(
+        triplet_first,
+        triplet_first + candidates.size(),
+        [func] __device__(auto val) { return func(thrust::get<2>(val)); },
+        handle.get_comms().get_size(),
+        std::numeric_limits<vertex_t>::max(),
+        handle.get_stream());
+
+      std::vector<size_t> h_tx_value_counts(d_tx_value_counts.size());
+      raft::update_host(h_tx_value_counts.data(),
+                        d_tx_value_counts.data(),
+                        d_tx_value_counts.size(),
+                        handle.get_stream());
+      handle.sync_stream();
+
+      std::forward_as_tuple(std::tie(candidates, offers_from_candidates, targets), std::ignore) =
+        shuffle_values(handle.get_comms(),
+                       thrust::make_zip_iterator(
+                         candidates.begin(), offers_from_candidates.begin(), targets.begin()),
+                       h_tx_value_counts,
+                       handle.get_stream());
```

Comment:
`shuffle_int_vertex_value_pairs_to_local_gpu_by_vertex_partitioning`?

Reply:
Here we are shuffling two values associated with one vertex, shuffle_int_vertex_value_pairs_to_local_gpu_by_vertex_partitioning only works for shuffling one value for each vertex.

Reply:
You mean shuffling a tuple associated with one vertex, right? Can't you update the existing function to support a tuple?

Reply:
```
candidates
offers_from_candidates 
 targets
```
Each target in `targets` has  a candidate and offer from the candidate. 

@seunghwak  Do you mean updating `shuffle_int_vertex_value_pairs_to_local_gpu_by_vertex_partitioning` to handle tuple of arbitrary length, and containing of any types?

Reply:
Yes, for the code (if this is not too complicated), no for explicit instantiation. For explicit instantiation, you can just add an explicit instantiation for your use case.

Reply:
Changed to groupby_gpu_id_and_shuffle_values